<p style="align: center;"><img align=center src="https://s8.hostingkartinok.com/uploads/images/2018/08/308b49fcfbc619d629fe4604bceb67ac.jpg" width=500 height=450/></p>

<h3 style="text-align: center;"><b>"Глубокое обучение". Продвинутый поток</b></h3>

<h2 style="text-align: center;"><b>День 3. Библиотека <a href="http://pandas.pydata.org/">pandas</a></b></h2>

Библиотека `pandas` активно используется в современном data science для работы с данными, которые могут быть представлены в виде таблиц (а это очень, очень большая часть данных)

`pandas` есть в пакете Anaconda, но если вдруг у Вас её по каким-то причинам нет, то можно установить, раскомментировав следующую команду:

In [0]:
# !pip install numpy pandas

In [0]:
import numpy as np
import pandas as pd

## pd.Series

Одномерный набор данных. Похож на `np.ndarray`, но с более расширенным функционалом

In [0]:
np.random.seed(42)

In [40]:
np_array = np.random.randint(low=-15, high=20, size=4)
np_array

array([13, -1, -8,  5])

In [41]:
# Series можно сделать из массива
pd_series_1 = pd.Series(np_array)
pd_series_1

0    13
1    -1
2    -8
3     5
dtype: int64

In [42]:
# Так же можно аннотировать индексы, чтобы потом было более удобно обращаться к элементам
pd_series_2 = pd.Series(np_array, index=['1st day', '2nd day', '3rd day', '4th day'])
pd_series_2

1st day    13
2nd day    -1
3rd day    -8
4th day     5
dtype: int64

In [43]:
pd_series_2['4th day']

5

In [44]:
# А еще можно дать pd.Series имя, чтобы было совсем красиво
pd_series_3 = pd.Series(np_array, index=['1st day', '2nd day', '3rd day', '4th day'], name='Temperature')
pd_series_3

1st day    13
2nd day    -1
3rd day    -8
4th day     5
Name: Temperature, dtype: int64

In [45]:
# с индексами можно работать так же, как и в случае с обычным list
print(pd_series_3[0])
print('*'*5)
print(pd_series_3[1:3])
print('*'*5)
print(pd_series_3[::-1])

13
*****
2nd day   -1
3rd day   -8
Name: Temperature, dtype: int64
*****
4th day     5
3rd day    -8
2nd day    -1
1st day    13
Name: Temperature, dtype: int64


`series` можно отсортировать как по значениям, так и по индексу

In [46]:
pd_series_3.sort_index()

1st day    13
2nd day    -1
3rd day    -8
4th day     5
Name: Temperature, dtype: int64

In [47]:
pd_series_3.sort_values()

3rd day    -8
2nd day    -1
4th day     5
1st day    13
Name: Temperature, dtype: int64

с `series` можно работать как с `np.ndarray`

In [48]:
pd_series_3 + 100

1st day    113
2nd day     99
3rd day     92
4th day    105
Name: Temperature, dtype: int64

In [49]:
np.exp(pd_series_3)

1st day    442413.392009
2nd day         0.367879
3rd day         0.000335
4th day       148.413159
Name: Temperature, dtype: float64

In [50]:
term_1 = pd.Series(np.random.randint(0, 10, 5))
term_2 = pd.Series(np.random.randint(0, 10, 6))

print('Term1:\n{}\n\nTerm2:\n{}\n\nSum:\n{}'.format(term_1, term_2, term_1 + term_2))

Term1:
0    6
1    9
2    2
3    6
4    7
dtype: int64

Term2:
0    4
1    3
2    7
3    7
4    2
5    5
dtype: int64

Sum:
0    10.0
1    12.0
2     9.0
3    13.0
4     9.0
5     NaN
dtype: float64


In [51]:
term_1.shape

(5,)

In [52]:
date_range = pd.date_range('20190101', periods=10)
pd_series_4 = pd.Series(np.random.rand(10), date_range)
pd_series_4

2019-01-01    0.056412
2019-01-02    0.721999
2019-01-03    0.938553
2019-01-04    0.000779
2019-01-05    0.992212
2019-01-06    0.617482
2019-01-07    0.611653
2019-01-08    0.007066
2019-01-09    0.023062
2019-01-10    0.524775
Freq: D, dtype: float64

In [53]:
pd_series_4 > 0.5

2019-01-01    False
2019-01-02     True
2019-01-03     True
2019-01-04    False
2019-01-05     True
2019-01-06     True
2019-01-07     True
2019-01-08    False
2019-01-09    False
2019-01-10     True
Freq: D, dtype: bool

в качестве индекса можно указать выражение, и нам будут возвращены только те элементы, для которых знвчение является `True`

In [54]:
pd_series_4[pd_series_4 > 0.5]

2019-01-02    0.721999
2019-01-03    0.938553
2019-01-05    0.992212
2019-01-06    0.617482
2019-01-07    0.611653
2019-01-10    0.524775
dtype: float64

In [55]:
pd_series_4[(pd_series_4 > 0.5) & (pd_series_4 < 0.6)]

2019-01-10    0.524775
Freq: D, dtype: float64

In [56]:
pd_series_4[(pd_series_4 > 0.6) | (pd_series_4 < 0.2)]

2019-01-01    0.056412
2019-01-02    0.721999
2019-01-03    0.938553
2019-01-04    0.000779
2019-01-05    0.992212
2019-01-06    0.617482
2019-01-07    0.611653
2019-01-08    0.007066
2019-01-09    0.023062
Freq: D, dtype: float64

## DataFrame

Двумерная таблица данных. Имеет индекс и набор столбцов (возможно, имеющих разные типы). Таблицу можно построить, например, из словаря, значениями в котором являются одномерные наборы данных.

In [57]:
# Dataframe можно составить из словаря. Ключ будет соответсовать колонке
d={'one':pd.Series([1,2,3],index=['a','b','c']),
   'two':pd.Series([1,2,3,4],index=['a','b','c','d'])}
df=pd.DataFrame(d)
df

one  two
a  1.0    1
b  2.0    2
c  3.0    3
d  NaN    4

In [58]:
df.index

Index(['a', 'b', 'c', 'd'], dtype='object')

In [59]:
df.columns

Index(['one', 'two'], dtype='object')

In [60]:
df['one']['c']

3.0

In [61]:
df.one

a    1.0
b    2.0
c    3.0
d    NaN
Name: one, dtype: float64

Если указать диапазон индексов, то это означает диапазон строк. Причём **последняя строка включается в таблицу.**

Диапазон целых чисел даёт диапазон строк с такими номерами, **не включая последнюю строку** (как обычно при индексировании списков). 

На это стоит обратить внимание

In [62]:
df['b':'d']

one  two
b  2.0    2
c  3.0    3
d  NaN    4

In [63]:
df[1:3]

one  two
b  2.0    2
c  3.0    3

In [64]:
df['three'] = df['one'] * df['two']
df

one  two  three
a  1.0    1    1.0
b  2.0    2    4.0
c  3.0    3    9.0
d  NaN    4    NaN

Иногда нам надо обьединить 2 таблицы вместе. В этом могут помочь методы `concat` и `join` 

In [65]:
# concatinating
df2=pd.DataFrame({'one':{'e':0,'f':1},'one_tr':{'e':2}})
df2

one  one_tr
e    0     2.0
f    1     NaN

In [66]:
pd.concat([df,df2], sort=False)

one  two  three  one_tr
a  1.0  1.0    1.0     NaN
b  2.0  2.0    4.0     NaN
c  3.0  3.0    9.0     NaN
d  NaN  4.0    NaN     NaN
e  0.0  NaN    NaN     2.0
f  1.0  NaN    NaN     NaN

In [67]:
df1 = pd.DataFrame({
    'one':pd.Series([1,2,3,4], index=['a','b','c','d']),
    'two':pd.Series([5,6,7,8], index=['a','b','c','d'])
})
df1

one  two
a    1    5
b    2    6
c    3    7
d    4    8

In [68]:

df2 = pd.DataFrame({
    'one':pd.Series([1,2,3,4,5], index=['a','b','c','d', 'e']),
    'three':pd.Series(['1','1','2','3'], index=['a','b','c','d'])
})
df2

one three
a    1     1
b    2     1
c    3     2
d    4     3
e    5   NaN

In [69]:
df1.join(df2.set_index('one'), on='one')

one  two three
a    1    5     1
b    2    6     1
c    3    7     2
d    4    8     3

Более подробно про join:  https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.join.html

Теперь попробуем поработать с настоящими данными

### Скачать данные: https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps

In [73]:
!wget https://www.dropbox.com/s/o9uiqxb3yr6ch6k/app-store-apple-data-set-10k-apps.zip?dl=0
!unzip app-store-apple-data-set-10k-apps.zip?dl=0

--2019-10-12 16:58:19--  https://www.dropbox.com/s/o9uiqxb3yr6ch6k/app-store-apple-data-set-10k-apps.zip?dl=0
Resolving www.dropbox.com (www.dropbox.com)... 162.125.1.1, 2620:100:601b:1::a27d:801
Connecting to www.dropbox.com (www.dropbox.com)|162.125.1.1|:443... connected.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: /s/raw/o9uiqxb3yr6ch6k/app-store-apple-data-set-10k-apps.zip [following]
--2019-10-12 16:58:19--  https://www.dropbox.com/s/raw/o9uiqxb3yr6ch6k/app-store-apple-data-set-10k-apps.zip
Reusing existing connection to www.dropbox.com:443.
HTTP request sent, awaiting response... 302 Found
Location: https://uc16f4058dc520467910f87b74a9.dl.dropboxusercontent.com/cd/0/inline/AqSQo3ZIGN1yw87yE7GYxqRxmwRp1_DPjBdgIwD1wX9Dm22_kgYQ25fJVQk80NP6PAdjB5I5zYK9fG6c9WqDllu44Gt7sfAru4wiQk0w_knr7npFYXofgRBf_1O_yRn1vgM/file# [following]
--2019-10-12 16:58:19--  https://uc16f4058dc520467910f87b74a9.dl.dropboxusercontent.com/cd/0/inline/AqSQo3ZIGN1yw87yE7GYxqRxmwRp1_DPjB

In [74]:
data = pd.read_csv('AppleStore.csv')

# вывести первые 5 строк
# data.head(5)
# вывести последние 5 строк
# data.tail(5)
# вывести 5 случайных строк
data.sample(5)

Unnamed: 0          id  ... lang.num  vpp_lic
5634        7638  1091186897  ...        1        1
4951        6457  1067024404  ...        1        1
4359        5504  1039910808  ...       13        1
1615        1874   572534490  ...       15        1
3554        4320   974529966  ...       11        1

[5 rows x 17 columns]

In [75]:
data = pd.read_csv('./AppleStore.csv', index_col=0)

data.head()

id  ... vpp_lic
1  281656475  ...       1
2  281796108  ...       1
3  281940292  ...       1
4  282614216  ...       1
5  282935706  ...       1

[5 rows x 16 columns]

Можно узнать размер таблицы, информацию о значениях таблицы, различные статистики по значениям

In [76]:
data.shape

(7197, 16)

In [77]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7197 entries, 1 to 11097
Data columns (total 16 columns):
id                  7197 non-null int64
track_name          7197 non-null object
size_bytes          7197 non-null int64
currency            7197 non-null object
price               7197 non-null float64
rating_count_tot    7197 non-null int64
rating_count_ver    7197 non-null int64
user_rating         7197 non-null float64
user_rating_ver     7197 non-null float64
ver                 7197 non-null object
cont_rating         7197 non-null object
prime_genre         7197 non-null object
sup_devices.num     7197 non-null int64
ipadSc_urls.num     7197 non-null int64
lang.num            7197 non-null int64
vpp_lic             7197 non-null int64
dtypes: float64(3), int64(8), object(5)
memory usage: 955.9+ KB


In [78]:
data.describe()

id    size_bytes  ...     lang.num      vpp_lic
count  7.197000e+03  7.197000e+03  ...  7197.000000  7197.000000
mean   8.631310e+08  1.991345e+08  ...     5.434903     0.993053
std    2.712368e+08  3.592069e+08  ...     7.919593     0.083066
min    2.816565e+08  5.898240e+05  ...     0.000000     0.000000
25%    6.000937e+08  4.692275e+07  ...     1.000000     1.000000
50%    9.781482e+08  9.715302e+07  ...     1.000000     1.000000
75%    1.082310e+09  1.819249e+08  ...     8.000000     1.000000
max    1.188376e+09  4.025970e+09  ...    75.000000     1.000000

[8 rows x 11 columns]

Чтобы посмотреть, какие колонки есть в таблице, можно воспользоваться `columns`

In [79]:
data.columns

Index(['id', 'track_name', 'size_bytes', 'currency', 'price',
       'rating_count_tot', 'rating_count_ver', 'user_rating',
       'user_rating_ver', 'ver', 'cont_rating', 'prime_genre',
       'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic'],
      dtype='object')

Метод `values` преобразует `pd.DataFrame` к `np.ndarray`

In [80]:
data.values

array([[281656475, 'PAC-MAN Premium', 100788224, ..., 5, 10, 1],
       [281796108, 'Evernote - stay organized', 158578688, ..., 5, 23, 1],
       [281940292, 'WeatherBug - Local Weather, Radar, Maps, Alerts',
        100524032, ..., 5, 3, 1],
       ...,
       [1187779532, 'Bret Michaels Emojis + Lyric Keyboard', 111322112,
        ..., 1, 1, 1],
       [1187838770, 'VR Roller Coaster World - Virtual Reality',
        97235968, ..., 0, 2, 1],
       [1188375727, 'Escape the Sweet Shop Series', 90898432, ..., 0, 2,
        1]], dtype=object)

In [81]:
type(data.values)

numpy.ndarray

In [82]:
data.values[:,0]

array([281656475, 281796108, 281940292, ..., 1187779532, 1187838770,
       1188375727], dtype=object)

Более продвинутое индексирование по таблице: **loc** и **iloc** 

`iloc` - index-based selection

In [83]:
data.iloc[0]

id                        281656475
track_name          PAC-MAN Premium
size_bytes                100788224
currency                        USD
price                          3.99
rating_count_tot              21292
rating_count_ver                 26
user_rating                       4
user_rating_ver                 4.5
ver                           6.3.5
cont_rating                      4+
prime_genre                   Games
sup_devices.num                  38
ipadSc_urls.num                   5
lang.num                         10
vpp_lic                           1
Name: 1, dtype: object

In [85]:
data.iloc[1:3, 1:3]

track_name  size_bytes
2                        Evernote - stay organized   158578688
3  WeatherBug - Local Weather, Radar, Maps, Alerts   100524032

In [86]:
data.iloc[[0,1,2], 0]

1    281656475
2    281796108
3    281940292
Name: id, dtype: int64

In [87]:
data.iloc[-1, :]

id                                    1188375727
track_name          Escape the Sweet Shop Series
size_bytes                              90898432
currency                                     USD
price                                          0
rating_count_tot                               3
rating_count_ver                               3
user_rating                                    5
user_rating_ver                                5
ver                                          1.0
cont_rating                                   4+
prime_genre                                Games
sup_devices.num                               40
ipadSc_urls.num                                0
lang.num                                       2
vpp_lic                                        1
Name: 11097, dtype: object

`loc` - label-based selection

In [88]:
# первое значение - index, второе - имя колонки
data.loc[1, ['id', 'track_name', 'price']]

id                  281656475
track_name    PAC-MAN Premium
price                    3.99
Name: 1, dtype: object

**Choosing between loc and iloc**

When choosing or transitioning between loc and iloc, there is one "gotcha" worth keeping in mind, which is that the two methods use slightly different indexing schemes.

`iloc` uses the Python stdlib indexing scheme, where the first element of the range is included and the last one excluded. So 0:10 will select entries 0,...,9. loc, meanwhile, indexes inclusively. So 0:10 will select entries 0,...,10.

Why the change? Remember that loc can index any stdlib type: strings, for example. If we have a DataFrame with index values `Apples, ..., Potatoes, ...,` and we want to select "all the alphabetical fruit choices between Apples and Potatoes", then it's a lot more convenient to index `df.loc['Apples':'Potatoes']` than it is to index something like `df.loc['Apples', 'Potatoet']` (t coming after s in the alphabet).

This is particularly confusing when the DataFrame index is a simple numerical list, e.g. 0,...,1000. In this case `df.iloc[0:1000]` will return 1000 entries, while `df.loc[0:1000]` return 1001 of them! To get 1000 elements using `loc`, you will need to go one lower and ask for `df.iloc[0:999]`.

Otherwise, the semantics of using `loc` are the same as those for `iloc`.

`loc` может принимать булевы выражения (так же, как и в случае с `pd.Series`)

Еще один полезный метод - `isin`

In [89]:
data.loc[data.prime_genre.isin(['Games', 'Shopping'])].head()

id  ... vpp_lic
1   281656475  ...       1
4   282614216  ...       1
6   283619399  ...       1
10  284736660  ...       1
11  284791396  ...       1

[5 rows x 16 columns]

Иногда бывает так, что мы хотим как-то сагрегировать данные, чтобы ответить на интересующие нас вопросы

Например, мы хотим узнать какая максимальная цена приложения, среди приложений, которые получили одинаковые оценки пользователей

для этого нам поможет метод `groupby`

In [90]:
data.groupby('user_rating').price.max().head()

user_rating
0.0    47.99
1.0    24.99
1.5     4.99
2.0     9.99
2.5     9.99
Name: price, dtype: float64

*Задания*

---

Получите ответ на следующие вопросы (с помощью pandas):

1). Сколько приложений с максимальным рейтингом (`user_rating`)?

2). Сколько всего жанров (`prime_genre`) и какие они? (Hint: `np.unique`)  

3). Сколько суммарно байт (`size_bytes`) весят все приложения жанра Finance?  

4). Сколько всего бесплатных приложений, у которых рейтинг юзеров больше трёх?  

5). Есть ли приложения, которые стоят больше 100 долларов?

6). Сколько приложений из жанра `Games` или `Shopping`, которые стоят дороже 10 USD или меньше 2 USD, но не являются бесплатными?

7). Посчитайте среднюю цену приложений, которые имеют рейтинг 4.5 и возрастное ограничение (`cont_rating`) 12+

---

Мы разобрали только малую часть работы с `pandas`. Это действительно мощный инструмент для работы с данными и его возможности обширны

Из того, что мы не разобрали (уделили мало внимания), стоит обратить внимание на следующие вещи:

* Работа с пропущенными значениями (`fillna`)
* Объединение таблиц (`join`)
* Агрегирование таблицы (`gropby`)
* Применение к данным функций (`apply`, `applymap`)
* так же стоит посмотреть, как работать с базами данных в `pandas`

Список полезных ссылок:

* Официальные туториалы: https://pandas.pydata.org/pandas-docs/stable/getting_started/tutorials.html

* Официальная документация: https://pandas.pydata.org/pandas-docs/stable/

* Статья на Хабре от OpenDataScience сообщества: https://habr.com/company/ods/blog/322626/

* Подробный гайд: https://media.readthedocs.org/pdf/pandasguide/latest/pandasguide.pdf

* Небольшой курс от kaggle: https://www.kaggle.com/learn/pandas

Главное в работе с новыми библиотеками -- не бояться тыкать в разные функции, смотреть типы возвращаемых объектов и активно пользоваться Яндексом, читать документацию на официальном сайте (она довольно хорошо написана), а ещё лучше понимать всё из docstring'а (`Shift+Tab` при нахождении курсора внутри скобок функции).